errors: display Node.js version at the end of stacktraces on fatal exception that causes exit

Divlo opened the pull request:
Display Node.js version at the end of stacktraces
on fatal exception that causes exit.
Easier for debugging so you don't have
to ask "what node version are you on?",
it is directly in the error the user
copy/paste from when asking for help.

Fixes: https://github.com/nodejs/node/issues/29731
<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

BridgeAR commented:
If we want to add the Node.js version, I would only do that to fatal exceptions that cause Node.js to end.

Divlo commented:
> If we want to add the Node.js version, I would only do that to fatal exceptions that cause Node.js to end.

Right, thanks for sharing your thoughts.
I don't really understand what you mean, because every errors exit the program.

Divlo commented:
Before continuing further the work on this PR, I think it's better to gather some feedbacks from the Node.js community and main contributors.

I made a file called `app.js` with this inside : 
```js
throw new Error("Some random error...")
```

Currently (with Node.js v16) by doing `node app.js`, I've got this output to `stderr` :
```
/home/divlo/Documents/testing/app.js:1
throw new Error("Some random error...")
^

Error: Some random error...
    at Object.<anonymous> (/home/divlo/Documents/testing/app.js:1:7)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Function.Module._load (node:internal/modules/cjs/loader:828:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
```

With the changes I made on this PR, here's the new output from the same file:

```
/home/divlo/Documents/testing/app.js:1
throw new Error("Some random error...")
^

Error: Some random error...
    at Object.<anonymous> (/home/divlo/Documents/testing/app.js:1:7)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Function.Module._load (node:internal/modules/cjs/loader:828:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
    via Node.js v17.0.0-pre
```

I see that's also bumping the major version number, is that considered as a **BREAKING CHANGE** or is it only when testing locally with the local build ?

Any thoughts ? 
Have you something to add @ORESoftware ?

aduh95 commented:
> I see that's also bumping the major version number

That's not bumping the major version number, the `master` branch is at version `v17.0.0-pre` so what you are seeing is expected. If you revert your commit and run `make -j12 && out/Release/node --version`, you'll see you're still at `v17.0.0-pre`.

> is that considered as a **BREAKING CHANGE**

https://github.com/nodejs/node/blob/005ebafbd17c8248447c2321269f508b3076ee1e/doc/guides/collaborator-guide.md#L341-L348

I don't think it qualifies as breaking change.



> > If we want to add the Node.js version, I would only do that to fatal exceptions that cause Node.js to end.
> 
> Right, thanks for sharing your thoughts.
> I don't really understand what you mean, because every errors exit the program (if don't try/catch them of course).

I think they meant that doing `console.log(new Error().stack)` should not out the Node.js version, it should only be added when the error is making the process crash (so when it's been thrown and not caught).

mhdawson commented:
@gireeshpunathil what's your thought on this one. I'm thinking it might be useful to have the additional info in the strack trace but I'm wondering if there are some down sides that I'm not thinking of?

Divlo commented:
Thanks for your concise answer! @aduh95 

Right @mhdawson, I'm happy to continue the work for this PR.
Could I have some help to fix the tests errors ?
I already tried to edit some tests to see what's going on but there is much more to fix, any insights, tips ?
Also, I make this PR ready for review, so the GitHub Actions are not skipped, so we can see what tests are failing. 
Concerning the commit message, I'll follow the Node.js commit convention when everything will be good to go and green, until then, I'll just commit, it makes life easier.

ORESoftware commented:
via Node.js version at <path> 

would be a nice touch, but I doubt node runtime has that info by default...maybe its process.argv[0]

gireeshpunathil commented:
@mhdawson , @Divlo  - 

any extra bit of info is definitely useful from the diagnostics point of view. The side effect of this I can think of are CI  and external tools - that may have static assumptions about the shape /structure of the stack trace.  I suggest to run CITGM tests too, once the CI passes - that can potentially give us hint around what can break and how.

fixing CI is easy, but bit laborious.
 - go to each failure
 - find the matching test case
 -  find what the expected value is
 - modify it based on the change that we made (mostly the expected is going to be a regexp, that needs to cater to the change too)

BridgeAR commented:
@Divlo if you change this to only be attached for fatal exceptions, you won't have to change as many tests. The last JS code that is run before exiting is: https://github.com/nodejs/node/blob/ba84524a7ccb7c668c4f19c958bb0c7dc5ab9cdf/lib/internal/process/execution.js#L170-L182 The stack trace itself is printed in the C++ layer: https://github.com/nodejs/node/blob/ba84524a7ccb7c668c4f19c958bb0c7dc5ab9cdf/src/node_errors.cc#L282 It should look similar to: https://github.com/nodejs/node/blob/ba84524a7ccb7c668c4f19c958bb0c7dc5ab9cdf/src/node_errors.cc#L396-L399

Divlo commented:
Following @BridgeAR recomendations, here's the new behavior (in a file called `errors.js`) :

<details>
<summary>Version should be printed</summary>

`errors.js` content :
```js
throw new Error('Some random error...')
```

`node errors.js` output :
```
/home/divlo/Documents/testing/errors.js:1
throw new Error('Some random error...')
^

Error: Some random error...
    at Object.<anonymous> (/home/divlo/Documents/testing/errors.js:1:7)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47
    via Node.js v17.0.0-pre
```
</details>

<details>
<summary>Version should not be printed because the error is in a try/catch</summary>

`errors.js` content :
```js
try {
  throw new Error('Some random error...')
} catch (error) {
  console.error(error)
}
```

`node errors.js` output :

```
Error: Some random error...
    at Object.<anonymous> (/home/divlo/Documents/testing/errors.js:4:9)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47
```
</details>

<details>
<summary>Version should be printed (in addition to <code>--trace-uncaught</code> warning)</summary>

`errors.js` content :
```js
throw {
  get stack() {
    throw new Error('weird throw but ok')
  },
  get name() {
    throw new Error('weird throw but ok')
  }
}
```

`node errors.js` output :

```
/home/divlo/Documents/testing/errors.js:9
throw {
^
[object Object]
(Use `node --trace-uncaught ...` to show where the exception was thrown)
    via Node.js v17.0.0-pre
```
</details>

Also good news, the CI now pass, everything should be green. :green_circle: 
I'm defintely up to run [CITGM](https://github.com/nodejs/citgm) tests, to see if there is some breakage.

Divlo commented:
Now that the CI pass, what is needed to move this PR forward ? @mhdawson @gireeshpunathil

aduh95 reviewed (approved):
Overall this LGTM, thanks for the hard work of fixing every stacktrace in the message tests. I'm a bit unsure if "via" is the correct word to use, but I also don't have a better suggestion :)
Just a few nits:

Divlo commented:
> @aduh95 I agree about the wording. What do you think about removing it including the indentation? Just the plain `Node.js vX.Y.Z`?

I don't mind about the wording, both looks good to me. :smile:
If I understand your feedback the error should be printed something like that :
```
/home/divlo/Documents/testing/app.js:1
throw new Error("Some random error...")
^

Error: Some random error...
    at Object.<anonymous> (/home/divlo/Documents/testing/app.js:1:7)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Function.Module._load (node:internal/modules/cjs/loader:828:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
Node.js v17.0.0-pre
```

Divlo commented:
rr

mhdawson reviewed (approved):
LGTM

devsnek commented:
i think we should definitely keep the indentation.

mhdawson commented:
While I agree this should be SemVer minor I wonder if it would be prudent to avoid backporting out of an abundance of caution as it does affect a good number of our own tests and may do the same for others as well. @nodejs/tsc  any thoughts on that?

targos commented:
If you think it shouldn't be backported, then it should be semver-major.

mhdawson commented:
I'd  be ok with marking it SemVer major even though technically its may not be.

richardlau commented:
While declaring something semver-major automatically excludes it from being backported, we can alternatively add the dont-land-on labels to achieve the same thing for things that are not semver-major but we still do not want to land on earlier release lines.

targos commented:
> While declaring something semver-major automatically excludes it from being backported, we can alternatively add the dont-land-on labels to achieve the same thing for things that are not semver-major but we still do not want to land on earlier release lines.

Sure, but if the reason for avoiding a backport is because we think the change can break someone, that's exactly what semver-major is for.

Divlo commented:
I think the best way to know how it affects the ecosystem and breakage is to run some [CITGM](https://github.com/nodejs/citgm) tests, depending on that, we could see if we should make it as semver-major or semver-minor.

Any instructions to how to run the tests ?

targos commented:
Here's a CITGM run: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2698/

mcollina reviewed (approved):
lgtm

mmarchini commented:
Now sure how effective a CITGM run would be in this case, since libraries/apps that would potentially break are more likely to be observability/debugging/introspection related. For example, Sentry has a [stacktrace parser](https://github.com/getsentry/sentry-javascript/blob/master/packages/node/src/stacktrace.ts) which I'm not entirely sure if it would be affected by the change or not.

FWIW I'm not opposed to this change but I do think it should be treated as semver-major.

richardlau commented:
It breaks at least tape's testsuite, e.g. https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2698/nodes=debian10-x64/testReport/junit/(root)/citgm/tape_v5_2_2/
```
         not ok 3 - should be equivalent
           ---
           found:
             - '$TEST/async-await/sync-error.js:7'
             - '    throw new Error(''oopsie'');'
             - '    ^'
             - ''
             - 'Error: oopsie'
             - '    at Test.myTest ($TEST/async-await/sync-error.js:$LINE:$COL)'
             - '    at Test.bound [as _cb] ($TAPE/lib/test.js:$LINE:$COL)'
             - '    at Test.run ($TAPE/lib/test.js:$LINE:$COL)'
             - '    at Test.bound [as run] ($TAPE/lib/test.js:$LINE:$COL)'
             - '    via Node.js v17.0.0-pre'
             - ''
           wanted:
             - '$TEST/async-await/sync-error.js:7'
             - '    throw new Error(''oopsie'');'
             - '    ^'
             - ''
             - 'Error: oopsie'
             - '    at Test.myTest ($TEST/async-await/sync-error.js:$LINE:$COL)'
             - '    at Test.bound [as _cb] ($TAPE/lib/test.js:$LINE:$COL)'
             - '    at Test.run ($TAPE/lib/test.js:$LINE:$COL)'
             - '    at Test.bound [as run] ($TAPE/lib/test.js:$LINE:$COL)'
             - ''
           at:
             line: 193
             column: 11
             file: test/async-await.js
           stack: |
             test/async-await.js:193:11
             ChildProcess.<anonymous> (test/common.js:102:9)
             ChildProcess.emit (node:events:394:28)
             Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
           source: |
             t.same(stripFullStack(stderr), [
           ...
```

Divlo commented:
Right, it is indeed a **BREAKING CHANGE**, we could probably include this in the next current release (v17).

mhdawson commented:
@Divlo I wonder if it would also be good to provide a command line option enable/disable as a work around for people/cases where it does break tests ?

Divlo commented:
> @Divlo I wonder if it would also be good to provide a command line option enable/disable as a work around for people/cases where it does break tests ?

@mhdawson The main idea of this change is to simplify debugging, so someone copy/pasting an error in stackoverflow for example, we can directly see what Node.js version is used.

@ORESoftware said (https://github.com/nodejs/node/issues/29731#issue-499177877) : 

> This could be added with a flag, and later turned off with a flag if you don't want to see it.

Maybe we can do that ? Now for the v16 release (in this PR), we add a new flag to enable this behavior (e.g `--enable-trace-version`), and for the next major release v17 we remove this flag and instead add a flag to disable it (e.g `--no-trace-version`, making it as the default behavior.
I think it is the right way to go, to have a smooth transition and less breaking while simplifying debugging and errors investigation.

targos commented:
I don't think a flag to enable the feature would be particularly useful.
If you can instruct the user to add it, it's easier to ask them to pass `--version` instead

Divlo commented:
> I don't think a flag to enable the feature would be particularly useful.
> If you can instruct the user to add it, it's easier to ask them to pass `--version` instead

Yeah, good point! The main idea of this PR is to avoid the need to ask the user to run `node --version`, when the user posted an error, by including this feature with a flag, it loses the whole point. :+1: @targos 

It's better to release it as a new major version, making it the behavior by default, but I still think we can include an option to disable it if the user really wants to.

mhdawson commented:
My thought was a flag to be able to disable it. ie on by default, but if it ends up causing a lot of breakage a way for people to work around that...

mhdawson commented:
My suggestion would be --disable-trace-version=<false|true> where the default is false and people could disable with --disable-trace-version=true if they needed to work around the breaking change for a period of time.

mmarchini commented:
I'm strongly +1 on having a flag to disable it (and having it enabled by default). This is great when pasting stacktraces to StackOverflow, but not so much for automated tooling that analyses stack traces.

I think for boolean flags we tend to use `--trace-version`/`--no-trace-version` syntax. Plus `--trace...` might lead to confusion with actual tracing features we provide. What about `--no-stacktrace-version`?

mmarchini reviewed (dismissed):
Making it explicit my opinion that there must be a way to disable this feature. Could be a flag or a runtime function/config. With the option to disable this feature I'm +1 on adding it.

danielleadams commented:
+1 if there is a way to disable with flag. `--no-stacktrace-version` sounds good. It has a similar format to other debugging flags such as `--no-deprecation`.

gengjiawen reviewed (changes requested):
Making it explicit my opinion that this feature should be disabled by default.

targos commented:
@gengjiawen The feature is only useful if it's enabled by default, though.

gengjiawen commented:
> @gengjiawen The feature is only useful if it's enabled by default, though.

Enabling this default is really weird.

aduh95 commented:
@gengjiawen can you elaborate on what you mean by "weird"? Are you saying this for current versions of Node.js specifically or in future versions as well?
 
Also, I agree that there's no value in having this feature if it's not enabled by default, as pointed out in https://github.com/nodejs/node/pull/38332#issuecomment-856929748.

gengjiawen commented:
> @gengjiawen can you elaborate on what you mean by "weird"? Are you saying this for current versions of Node.js specifically or in future versions as well?
> 
> Also, I agree that there's no value in having this feature if it's not enabled by default, as pointed out in [#38332 (comment)](https://github.com/nodejs/node/pull/38332#issuecomment-856929748).

I have not seen other language runtime print their version in stacktraces. It's hard for me to think developers not knowing this.

targos commented:
With Node.js, people who use versions managers like `nvm` or `Volta` often have many different versions installed on their machines. It's easy to run some code with the wrong one by mistake. I know that seeing the version number in the stack trace would be very useful in those cases because you can see immediately that it's not the right one.

mhdawson commented:
I'm ok with the alternate suggestions for the command line options.
Why I prefer the =<true|false> style though is that it makes it easier if we need to change the default.

For example if we use --no-stacktrace-version without <true|false> but find there is too much breakage then we might need to scramble to revert or disable by default and add another command line to enable. If we have an option which is --no-stacktrace-version=<true|false> then we can just flip the default which will likely be a 1 line commit with little/no risk.

targos commented:
I don't know if our options parser supports it, but in V8 for example, the boolean flags don't work with `--flag=true|false` but with `--flag` for true and `--no-flag` for false, and it's still a one-line change to flip the default.

mhdawson commented:
@targos I don't know either but if we can easily support --flag and --no-flag that sounds good to me.

mmarchini commented:
I was also thinking `--flag`/`--no-flag`, it's a pretty common CLI options pattern.

Divlo commented:
I just added the `--no-stacktrace-version` cli option. :+1: 
Should we also add the `--stacktrace-version` option ?
I didn't write tests for this new option yet, where and how should I test this new behavior?

> I have not seen other language runtime print their version in stacktraces. It's hard for me to think developers not knowing this.

@gengjiawen It is not because others runtime don't do it, Node.js should not do it, in contrary, fortunately, tools and runtimes compete together and work differently, and every runtime try to improve the user experience.
Maybe later, others runtime will follow because Node.js did it (I'm thinking of `cpython` where it might be useful too).

mhdawson commented:
In terms of tests you can add the option in a test with something like

// Flags: --no-stacktrace-version

in the test file. This is a test that adds a command line option: https://github.com/nodejs/node/blob/master/test/parallel/test-async-hooks-destroy-on-gc.js as an example.

In terms of --stacktrace-version I'd prefer if we added it now as well.

Divlo commented on test/parallel/test-no-stacktrace-version.js:
```diff
@@ -0,0 +1,27 @@
+    assert.ok(stderr.split('\n').length);
+
+    // Proxy the args for more tests.
+    callback(err, stdout, stderr);
```
I copy/paste this function from https://github.com/nodejs/node/blob/fa1a842a08aa0be1cbc7b6a0e5517340c6768986/test/parallel/test-error-reporting.js#L28-L40

The only change is the `cmd` variable to pass the right CLI option, there should be a way to follow the DRY principle, how could we make it reusable, so we could use it in `test/parallel/test-no-stacktrace-version.js`, `test/parallel/test-stacktrace-version.js` and `test/parallel/test-error-reporting.js` ?

Divlo commented:
Thanks for your help! @mhdawson
Done, I added both `--stacktrace-version` and `--no-stacktrace-version` options to the CLI.
By default we don't need to pass the `--stacktrace-version` option, because it will be taken by default (true).

Also I added 2 test file inside `test/parallel` folder, I'm not sure if there is a better way to write these tests (see my comment).

targos commented:
@addaleax Do you think it would be easily doable to change the options parser to automatically support `--boolean` and `--no-boolean` like in V8?

addaleax commented:
@targos Yes. Uncompiled, untested diff in the fold:

<details>

```diff
diff --git a/src/node_options-inl.h b/src/node_options-inl.h
index 4e1a12296bc7..ef54bb0caea6 100644
--- a/src/node_options-inl.h
+++ b/src/node_options-inl.h
@@ -225,6 +225,10 @@ inline std::string RequiresArgumentErr(const std::string& arg) {
   return arg + " requires an argument";
 }
 
+inline std::string NegationImpliesBooleanError(const std::string& arg) {
+  return arg + " is an invalid negation because it is not a boolean option";
+}
+
 // We store some of the basic information around a single Parse call inside
 // this struct, to separate storage of command line arguments and their
 // handling. In particular, this makes it easier to introduce 'synthetic'
@@ -325,6 +329,13 @@ void OptionsParser<Options>::Parse(
         name[i] = '-';
     }
 
+    // Convert --no-foo to --foo and keep in mind that we're negating.
+    bool is_negation = false;
+    if (name.find("--no-") == 0) {
+      name.erase(2, 3); // remove no-
+      is_negation = true;
+    }
+
     {
       auto it = aliases_.end();
       // Expand aliases:
@@ -410,9 +421,13 @@ void OptionsParser<Options>::Parse(
       }
     }
 
+    if (is_negation && info.type != kBoolean) {
+      errors->push_back(NegationImpliesBooleanError(arg));
+    }
+
     switch (info.type) {
       case kBoolean:
-        *Lookup<bool>(info.field, options) = true;
+        *Lookup<bool>(info.field, options) = !is_negation;
         break;
       case kInteger:
         *Lookup<int64_t>(info.field, options) = std::atoll(value.c_str());
```

</details>

mhdawson commented:
@targos are you thinking that should be added as part of this PR or that the PR can land as is and adding the general functionality to support the --boolean and --no-boolean would be added later?

targos commented:
I opened https://github.com/nodejs/node/pull/39023 for the general functionality

mhdawson commented:
@targos thanks for the update, makes sense to wait for that.

targos commented:
@Divlo https://github.com/nodejs/node/pull/39023 just landed. If you rebase on `master`, you can define the option only once with a `true`, and document it as a negation (like `--no-warnings` for example).

Divlo commented:
Done, thanks for the PR! @targos 
It should be good now.

gireeshpunathil commented:
this can have some reviews

mcollina reviewed (approved):
lgtm

mhdawson reviewed (approved):
Updated version still LGTM

mhdawson commented:
@gengjiawen I see there was some discussion with respect to your objection. I'm wondering if your objection is still blocking?

gengjiawen commented:
> @gengjiawen I see there was some discussion with respect to your objection. I'm wondering if your objection is still blocking?

Yeap. I still want this not enabled by default.

gireeshpunathil commented:
> Yeap. I still want this not enabled by default.

@gengjiawen - can you please elaborate your reason for the same? With the flags incorporated, now the only question is: what should be the default - on or off.

Given the background of this PR, its intended purpose is met only when the default is on, and that is driving the default "on" argument. what is the explanation for default "off" argument?

From the thread, I see two points: i) may break user land, ii) can always switch-on for those who want it. I guess both of those are answered / address.

Are there other points, or further explanations to the above two?

gengjiawen commented:
> what is the explanation for default "off" argument?

It's simply not worth it to turn it on. Besides the two points you pointed out: 

* we have report module https://nodejs.org/dist/latest-v16.x/docs/api/report.html with much more detailed info (It auto generate when Node.js process crashes).
* I think knowing your Node.js version you are running is easy enough for everyone.

gireeshpunathil commented:
@gengjiawen - thanks for the quick explanation

> we have report module https://nodejs.org/dist/latest-v16.x/docs/api/report.html with much more detailed info (It auto generate when Node.js process crashes).

the report feature is on by default, but a report is not produced by default on exceptions

> I think knowing your Node.js version you are running is easy enough for everyone.

agree, but the question is, how does one know the version information without needing to run / inspect the execution environment again. It is useful for those who help / support user problems and reduces one round-trip.

so while running `node -v` is trivial and easy, this feature makes it even simpler.

what do you think?

gengjiawen commented:
> so while running node -v is trivial and easy, this feature makes it even simpler.

Thanks for sharing the thought. 
But I still think it's not worth it. It's the same for python and ruby too.
I think the weighing behind this it will cause more trouble than the benefit.

```console
$ python test.py
Traceback (most recent call last):
  File "test.py", line 1, in <module>
    printt('a')
NameError: name 'printt' is not defined
```

Maybe we can bring a more common flag to fix this. With that flag, we provide much more comprehensive information ?

BridgeAR commented on test/message/error_aggregateTwoErrors.out:
```diff
@@ -11,3 +11,4 @@ AggregateError: original
     at node:internal/main/run_main_module:*:* {
   code: 'ERR0'
 }
+    via Node.js *
```
This looks rather confusing that it's indented after the closing bracket. Thus, I would not do that. It's not that important though.

Divlo replied:
Agreed.
Updated the PR without indentation and "via" keyword.
See: https://github.com/nodejs/node/pull/38332#issuecomment-910089818

BridgeAR reviewed (approved):
While not having a strong opinion about it, I can't see an actual downside and think it's fine to add this.

tniessen commented:
I don't have much of an opinion on this, I just want to throw an idea in based on stack traces I have seen in other environments. Because users cannot really rely on internal file names, we could preface internal module names, which do usually appear in stack traces, with `node:node-vX.Y.Z/` instead of `node:`.

```
Error
    at Timeout._onTimeout ([eval]:1:26)
    at listOnTimeout (node:internal/timers:557:17)
    at processTimers (node:internal/timers:500:7)
```

could be changed to

```
Error
    at Timeout._onTimeout ([eval]:1:26)
    at listOnTimeout (node:node-v16.6.0/internal/timers:557:17)
    at processTimers (node:node-v16.6.0/internal/timers:500:7)
```

richardlau commented:
> > Here's a CITGM run: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2698/
> 
> Quoting this previous comment here because I wrongly stated in the TSC meeting today that CITGM had not been run yet and I would run it. Whoops, yes, it has been run. (Thanks, Gireesh and Michaël!)

And in case anyone missed the follow up to that the CITGM run at the time showed this broke tape's testsuite: https://github.com/nodejs/node/pull/38332#issuecomment-856308452

apapirovski commented:
My 2c is the `via` with indentation is pretty confusing. I would second @aduh95 & @BridgeAR re: removing indentation and keeping it left-aligned. Also makes it clear it's not related to the stack trace.

@gireeshpunathil is there other diagnostic information we could print here that would be generally helpful in your opinion? If we're making this change would be good to consider that as well and just get it all done in one go.

targos reviewed (changes requested):
Just making sure the conversation in https://github.com/nodejs/node/pull/38332#pullrequestreview-739497393 is resolved before we merge.

gireeshpunathil commented:
> @gireeshpunathil is there other diagnostic information we could print here that would be generally helpful in your opinion? If we're making this change would be good to consider that as well and just get it all done in one go.

@apapirovski - idealy there are bunch of information that would come handy in a first-failure data capture, however, many info are relevant only based the problem at hand - for example, if the issue is related to memory, JavaScript heap info may be useful, etc.

The common useful info (that usually users provide partially or differently and would benefit if we get along side the stack trace) are `osName, osVersion, osMachine (arch)` fields in a diagnostic report. But I wouldn't term those as mandatory for this PR.

Divlo commented:
PR updated, here are the changes :
- We now use the flag called `--no-extra-info-on-fatal-exception` instead of `--no-stacktrace-version`.
- Instead of printing the Node.js version with the indentation and the "via" keyword, it is separated from the error stacktrace.
That means the error is now printed like that:

`errors.js` content :
```js
throw new Error('Some random error...')
```

`node errors.js` output :
```
/home/divlo/Documents/testing/errors.js:1
throw new Error('Some random error...')
^

Error: Some random error...
    at Object.<anonymous> (/home/divlo/Documents/testing/errors.js:1:7)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47

Node.js v17.0.0-pre
```

> My 2c is the via with indentation is pretty confusing. I would second @aduh95 & @BridgeAR re: removing indentation and keeping it left-aligned. Also makes it clear it's not related to the stack trace. And "via" just seems like a weird word to use here so would cut it.

I agree with @apapirovski, @aduh95 and @BridgeAR.
I think it is better without indentation and "via" keyword, that's why I made this change, but if not wanted, I could revert this change and make it print like before of course.

It could be cool, if this PR gets again a few reviews with the changes made, thanks a lot for your patience and your help! :smile:

targos commented on doc/api/cli.md:
```diff
@@ -1111,6 +1111,13 @@ added: v6.0.0
 
 Print stack traces for process warnings (including deprecations).
 
+### `--no-extra-info-on-fatal-exception`
```
this list is sorted alphabetically. should be just after `--no-deprecation`

Divlo replied:
Done, updated! :+1:

targos commented on doc/node.1:
```diff
@@ -449,6 +449,9 @@ Enabling this option may affect garbage collection behavior negatively.
 .It Fl -trace-warnings
 Print stack traces for process warnings (including deprecations).
 .
+.It Fl -no-extra-info-on-fatal-exception
```
ditto

aduh95 commented on test/parallel/test-no-extra-info-on-fatal-exception.js:
```diff
@@ -0,0 +1,27 @@
+}
+
+errExec('throws_error.js', common.mustCall((err, stdout, stderr) => {
+  assert.match(stderr, /at node:internal\/main\/run_main_module:17:47\r?\n$/);
```
Would it make more sense to add that test in `test/message` instead? This will break as soon as the `run_main_module` is updated.

Divlo replied:
I think, it would make more sense to add this small single test in `test/parallel/test-error-reporting.js` file so we avoid the duplication of the method `errExec` as mentioned in https://github.com/nodejs/node/pull/38332#discussion_r650089711, what do you think?

aduh95 replied:
That would work too but it looks to me that `test/message` would still be the right place for this test:

> Tests in `test/message` pass if the output matches the expected output.
> Message tests are particularly useful when checking for exact error messages.

You just need to create `test/message/test-no-extra-info-on-fatal-exception.js` which contains:

```js
// Flags: --no-extra-info-on-fatal-exception
require('../common');

throw new Error('foo');
```

And then you need to create `test/message/test-no-extra-info-on-fatal-exception.out` which contains:

```
*:3
throw new Error('foo');
^

Error: foo
    at Object.<anonymous> (*:3:7)
    at Module._compile (node:internal/modules/cjs/loader:*:*)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:*:*)
    at Module.load (node:internal/modules/cjs/loader:*:*)
    at Function.Module._load (node:internal/modules/cjs/loader:*:*)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:*:*)
    at node:internal/main/run_main_module:*:*
```

If you prefer to put it in `test/parallel/test-error-reporting.js`, you need to make the test not dependent on the precise line number and column number for `node:internal/main/run_main_module` (using `:\d+:\d+` instead of `:17:47`).

Divlo replied:
Indeed your first solution looks cleaner! :smile:
Done, updated!

aduh95 commented on test/parallel/test-error-reporting.js:
```diff
@@ -77,6 +77,5 @@ errExec('throws_error6.js', common.mustCall((err, stdout, stderr) => {
 errExec('throws_error7.js', common.mustCall((err, stdout, stderr) => {
-  assert.match(stderr,
-               /throw {\r?\n\^\r?\n{ toString: \[Function: toString] }\r?\n$/);
+  assert.match(stderr, /throw {\r?\n\^\r?\n{ toString: \[Function: toString] }\r?\n\nNode\.js \S+\r?\n$/);
```
```suggestion
  assert.match(stderr, /throw {\r?\n\^\r?\n{ toString: \[Function: toString] }\r?\n\r?\nNode\.js \S+\r?\n$/);
```

Divlo commented:
The CI keeps failing.
Is there is anything needed to make it pass or it is not on the side of this PR?

targos commented:
It fails because of this error (I think): 

```
13:23:03 ERROR: The process with PID 35016 (child process of PID 42664) could not be terminated.
```

/cc @nodejs/build-infra

mhdawson commented:
Landed as 59d3d542d69f078cd1d9f50f26b83944c381724e

mhdawson commented:
@Divlo thanks for you work and patience on this one.
